### PR TITLE
Provide allocation UUIDs for IP addresses.

### PR DESCRIPTION
**Pull request by Gerrit91**
This is required for accounting as otherwise we will not be able to distinguish if an IP address was re-allocated or not.

The database requires a migration for this as we started to discuss in #69 because the allocation UUID needs to be generated for all "old" IP address entities in the database.

### CODE REVIEW

File: cmd/metal-api/internal/service/machine-service.go [hunk 1]
```diff
@@ -1311,12 +1312,18 @@ func gatherUnderlayNetwork(ds *datastore.RethinkStore, allocationSpec *machineAl
...
+			return nil, fmt.Errorf("unable to create uuid for IP allocation: %v", err)
+		}
+
 		ipAddress, ipParentCidr, err := allocateIP(n.network, "", ipamer)
 		if err != nil {
 			return nil, fmt.Errorf("unable to allocate an ip in network: %s %#v", n.network.ID, err)
 		}
 		ip := &metal.IP{
 			IPAddress:        ipAddress,
+			AllocationUUID:   allocationUUID.String(),
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go:1326]: This is not the only place where a IP is allocated, adding the allocationUUID must be done in a central place

Comment (Gerrit91) [reply cmd/metal-api/internal/service/machine-service.go:1326]: Mh, yes, that would be better. But it's also a bigger refactoring to make this look adequate... I am not sure, if this refactoring should be contained in this PR.

The IP create function of the IP service and the machine network auto ip allocation should be able to use a common `ipCreate` function that hands over an IPSpec as a parameter, I guess.

Comment (majst01) [reply cmd/metal-api/internal/service/machine-service.go:1326]: TBH, this PR looks crazy as hell

**Comment by Gerrit91** [comment]
I wrote down a couple of lines that would enable us to run zero-downtime migrations of the RethinkDB sufficient for most of the update scenarios we did encounter in the past. I did it this way because there are no existing libraries available for RethinkDB, which would do such a thing and providing such a mechanism is a little different for RethinkDB than it is for usual SQL databases. You need to keep in mind that RethinkDB does not have a schema and that the model is actually described by the database clients, there are no transactions or things like that and we have multiple instances of clients running at the same time.

Therefore, the idea was that the deployment runs a post-upgrade hook that migrates the database to the latest state after new pods were rolled out. Reason that it is run after the deployment is that old clients should not be writing old models into the database after the migration has run. Of course there is one downside of running the migration after the rollout, which is that for a short period of time a client will not return the migrated data from the database, which is "kind-of" acceptable considering that the consumers of the metal-api need to update to the new version in most scenarios.

Summary of the approach:
- Add a "migration" table that holds the current migration version of the database
- A `migrate` command triggers the migration and does a migration when a newer version is available
- The pods of our metal-api replica set do not use the admin user anymore, such that the database can be put into read-only mode (this helps to avoid possible modification / concurrency issues during migration tasks)

Nevertheless, there can be database changes that require a downtime (e.g. changing a type of an existing field). In this case we would need to cut off client connections if we do not want to return with internal crashes to the clients until the migration has run... this would need to be done manually or require more sophisticated deployment strategies, which can determine if downtime is necessary and scale down / up the metal-api replica set for the migration. At the moment, we are not in this situation that we need such an upgrade path and I think that most of the time we can get away with this simple mechanism instead. It will certainly require some elaboration when we come into such a situation.

Maybe we should also put the "SemVer" migration that we had in the past into the list of migrations. Then we would support updating from metal-api v0.6.x to v0.7.x and later, which is not the case at hte moent and could lead to weird issues for someone who attempts such a version upgrade.

**Comment by Gerrit91** [comment]
Of course I am open for other ideas. This actually did not took so much time to implement (few hours). If there are better ideas, please tell me. It's just that I do not want to put in a "temporary migration endpoint" again for this because I think that breaks the update experience.

File: cmd/metal-api/internal/service/machine-service.go (hunk 1 above)
Comment (Gerrit91) [reply cmd/metal-api/internal/service/machine-service.go:1326]: I can't deny that. But it looked simpler to me than other approaches in regards of time and suitability. I am open for ideas and comments. :)

**Comment by Gerrit91** [comment]
@ulrichSchreiner We already talked about this a little in the background. What do you think about this? Would this be mergeable from your point of you or is it too "crazy as hell"?

**Comment by majst01** [comment]
First question: how can read-only replicas ever gets back into read-write mode after the upgrade was done ?

**Comment by Gerrit91** [comment]
It's done dynamically via the "grant" command of rethinkDB: https://rethinkdb.com/api/javascript/grant

The read-only mode should only be activated for a very short amount of time, just as long as the migration takes.

**Comment by majst01** [comment]
Probably replaced partly with: https://github.com/metal-stack/go-ipam/pull/28

**Comment by Gerrit91** [comment]
As this PR is likely not to be merged because the IP UUIDs will probably become part of a newer go-ipam version, I am closing this in favor of #98.

**Pull request closed without merging**